staging/src/k8s.io/apiserver/pkg/cel/types.go

```diff
@@ -440,37 +442,41 @@ func (rt *DeclTypeProvider) FindDeclType(typeName string) (*DeclType, bool) {
 	return rt.findDeclType(typeName)
 }
 
-// FindFieldType returns a field type given a type name and field name, if found.
+// FindStructFieldNames returns the field names associated with the type, if the type
+// is found.
+func (rt *DeclTypeProvider) FindStructFieldNames(typeName string) ([]string, bool) {
+	return []string{}, false
+}
+
+// FindStructFieldType returns a field type given a type name and field name, if found.
 //
 // Note, the type name for an Open API Schema type is likely to be its qualified object path.
 // If, in the future an object instance rather than a type name were provided, the field
 // resolution might more accurately reflect the expected type model. However, in this case
 // concessions were made to align with the existing CEL interfaces.
-func (rt *DeclTypeProvider) FindFieldType(typeName, fieldName string) (*ref.FieldType, bool) {
+func (rt *DeclTypeProvider) FindStructFieldType(typeName, fieldName string) (*types.FieldType, bool) {
 	st, found := rt.findDeclType(typeName)
 	if !found {
-		return rt.typeProvider.FindFieldType(typeName, fieldName)
+		return rt.typeProvider.FindStructFieldType(typeName, fieldName)
 	}
 
 	f, found := st.Fields[fieldName]
+	if rt.recognizeKeywordAsFieldName && !found && celReservedSymbols.Has(fieldName) {
```

Comment:
follow-up: do we need a unit test that ensures celReservedSymbols never gets something added to it that contains __, since that would need escaping? in a case like that, we'd probably pre-bake a map of `reserved symbol → escaped reserved symbol` rather than constructing by prefixing+suffixing with `__`, etc

Reply:
`celReservedSymbols` is currently hard coded in https://github.com/kubernetes/kubernetes/blob/ad805387e26a4e76ab53505d5eb552686368180b/staging/src/k8s.io/apiserver/pkg/cel/escaping.go#L25-L34 and the list is from https://github.com/google/cel-spec/blob/master/doc/langdef.md#syntax
Agree it would be great to have guard to prevent new value accidentally mess with existing escaping, the issue is tracked: https://github.com/kubernetes/kubernetes/issues/126311